### ISSUE TITLE
[Feature Request] Add Support For WireGuard Over Shadowsocks (or other obfuscation protocols)

### ISSUE THREAD

**Issue by ProgressiveArchitect** [mullvad/mullvadvpn-app#1049]
Currently, there is in-app support for Shadowsocks being used as multi-hop to connect to OpenVPN servers. I'd like to see the same support available when connecting to WireGuard servers.

**Comment by faern** [comment]
This is blocked on the fact that WireGuard currently can't operate over TCP and Shadowsocks can't really relay UDP. But yes, using bridges with WireGuard connections is surely a desired feature, so it's something we will work on when there is time.

**Comment by faern** [comment]
Let's keep the issue open. It's something we want to implement. If it's open other people might see it easier and not create duplicate issues.

**Comment by maximbaz** [comment]
> This is blocked on the fact that WireGuard currently can't operate over TCP and Shadowsocks can't really relay UDP

According to [the usage page](https://github.com/shadowsocks/shadowsocks-libev#usage), Shadowsocks does support relaying UDP, see `-u (Enable UDP relay)` and `-U (Enable UDP relay and disable TCP relay)`.

Does this mean this could potentially be implemented now, or there is more to this?

**Comment by PacoBell** [comment]
@faern Any comment on the apparent removal of this blocker? I've seen evidence of folks getting this working on their own single VPS instances. Someone just needs to code the backend now to glue it all together.

**Comment by ProgressiveArchitect** [comment]
@faern It looks like ShadowSocks can be routed over UDP now. So any updates for implementing this on WireGuard?

**Comment by faern** [comment]
Last time I checked it was not using purely UDP. Even if you enable the UDP relaying it was handshaking over TCP. But that might have changed. Please also note that we use the `shadowsocks-rs` Rust implementation in this VPN app, so it has to be supported there for us to use it. Changing which Shadowsocks implementation we use would be a bigger task.

**Comment by aveao** [comment]
A year later, shadowsocks-rs has a "udp_only" mode and it's possible to relay wireguard traffic through it.

So, any updates?

**Comment by faern** [comment]
We are currently working looking at ways of tunneling WireGuard etc. But we are currently not looking at shadowsocks. Thanks for the update on their UDP only support.

**Comment by whywhah** [comment]
It's been another year and Wireguard still has no Shadowsocks or other UDP tunneling. Just a reminder that this feature is still pretty much needed.

**Comment by ProgressiveArchitect** [comment]
@faern I've noticed your language shift from "shadowsocks" to "obfuscation". Does this mean that the Mullvad devs intend to utilize a different obfuscation protocol? If so, which protocol is being looked at as most likely to be implemented with UDP in mind?

**Comment by faern** [comment]
We shift the language because our obfuscation support is not only for Shadowsocks. We recently merged (#3431) a new "obfuscation engine" or whatever you want to call it, for our WireGuard connections. This is a framework in `mullvad-daemon` that allows it to connect WireGuard over any proxy implementation that can listen to a localhost UDP port and in some way send that obfuscated over the network.

Currently the only supported protocol is `udp-over-tcp`. This allows connecting to WireGuard servers using TCP. We supported this before the mentioned PR, but then we modeled it differently internally, now it's classified as an obfuscation protocol instead.

Shadowsocks is likely going to be the next protocol added to this new obfuscation framework of ours. But it's not 100% decided upon yet.

You configure what obfuscation WireGuard should use via the CLI command `mullvad obfuscation`. This has not been released yet, but if you build from the latest source code you can see this subcommand.

**Comment by PragmaTwice** [comment]
Hi, any updates here?

I want to try mullvad but I am live in a country where bridge mode is needed in my phone.

Bridge mode on phone app will hugely facilitate many people like me and make mallvad a more good choice.

**Comment by faern** [comment]
Hi @PragmaTwice.

WireGuard over TCP has been in the desktop app for a while now. Not yet on mobile, it's in the backlog.
We are also looking at enabling custom proxies for WireGuard for desktop. But nothing on using our own bridges yet. However, this is more at the idea stage than implementation stage so far sadly.

**Comment by Pilaton** [comment]
Is there any news?

**Comment by faern** [comment]
Not really. What I can say is that it's being frequently discussed as an anti-censorship measure we want to add. But it's not currently at the top of the pile. So other anti-censorship measures will be implemented during Q3.

**Comment by Cohenl19** [comment]
Hey any updates? It's 2024 and still wondering if there are ways to use Wireguard let's say at a school or library that would otherwise block the VPN connection. I've used shadowsocks before and I liked it, I just wasn't  much of a fan of needing to use OpenVPN

**Comment by duxsco** [comment]
fyi, the release [android/2024.5](https://github.com/mullvad/mullvadvpn-app/releases/tag/android%2F2024.5) features Wireguard over Shadowsocks, but doesn't mention whether UDP is used. At least, there are Shadowsocks bridges that run over UDP:

```
❯ curl -fsS https://api.mullvad.net/app/v1/relays | jq .bridge.shadowsocks
[
  {
    "protocol": "tcp",
    "port": 443,
    "cipher": "aes-256-gcm",
    "password": "mullvad"
  },
  {
    "protocol": "udp",
    "port": 1234,
    "cipher": "aes-256-cfb",
    "password": "mullvad"
  },
  {
    "protocol": "udp",
    "port": 1236,
    "cipher": "aes-256-gcm",
    "password": "mullvad"
  }
]
```

**Comment by duxsco** [comment]
Now, I use version 2024.6-beta2 on Linux, and Wireguard with obfuscation via Shadowsocks over UDP works.

**Comment by Serock3** [comment]
### :tada: :ninja:  [WireGuard over Shadowsocks has been released for desktop and Android](https://mullvad.net/en/blog/introducing-shadowsocks-obfuscation-for-wireguard) :ninja: :tada: 
It's coming to iOS sometime later, I'll close this issue when that happens.

**Comment by Serock3** [comment]
The feature is now available on all platforms, including iOS.
